pkg/storegateway/indexheader/reader_pool_test.go

```diff
@@ -28,54 +28,108 @@ import (
 
 func TestReaderPool_NewBinaryReader(t *testing.T) {
 	tests := map[string]struct {
-		lazyReaderEnabled     bool
-		lazyReaderIdleTimeout time.Duration
+		lazyReaderEnabled                           bool
+		lazyReaderIdleTimeout                       time.Duration
+		eagerLoadReaderEnabled                      bool
+		initialSync                                 bool
+		createLazyLoadedHeadersSnapshotFn           func(blockId ulid.ULID) lazyLoadedHeadersSnapshot
+		expectedLoadCountMetricBeforeLabelNamesCall int
+		expectedLoadCountMetricAfterLabelNamesCall  int
 	}{
 		"lazy reader is disabled": {
-			lazyReaderEnabled: false,
+			lazyReaderEnabled:                          false,
+			expectedLoadCountMetricAfterLabelNamesCall: 0, // no lazy loading
 		},
 		"lazy reader is enabled but close on idle timeout is disabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: 0,
+			lazyReaderEnabled:                          true,
+			lazyReaderIdleTimeout:                      0,
+			expectedLoadCountMetricAfterLabelNamesCall: 1,
 		},
 		"lazy reader and close on idle timeout are both enabled": {
-			lazyReaderEnabled:     true,
-			lazyReaderIdleTimeout: time.Minute,
+			lazyReaderEnabled:                          true,
+			lazyReaderIdleTimeout:                      time.Minute,
+			expectedLoadCountMetricAfterLabelNamesCall: 1,
+		},
+		"block is present in pre-shutdown loaded blocks and eager-loading is enabled during initial sync": {
+			lazyReaderEnabled:      true,
+			lazyReaderIdleTimeout:  time.Minute,
+			eagerLoadReaderEnabled: true,
+			initialSync:            true,
+			expectedLoadCountMetricBeforeLabelNamesCall: 1, // the index header will be eagerly loaded before the operation
+			expectedLoadCountMetricAfterLabelNamesCall:  1,
+			createLazyLoadedHeadersSnapshotFn: func(blockId ulid.ULID) lazyLoadedHeadersSnapshot {
+				return lazyLoadedHeadersSnapshot{
+					IndexHeaderLastUsedTime: map[ulid.ULID]int64{blockId: time.Now().UnixMilli()},
+					UserID:                  "anonymous",
+				}
+			},
+		},
+		"block is present in pre-shutdown loaded blocks and eager-loading is enabled after initial sync": {
```

Comment:
[nit] Similar to above
```suggestion
		"block is present in pre-shutdown loaded blocks and eager-loading is enabled, loading index header after initial sync": {
```